### DISCUSSION TITLE
Correct way to allocate array memory and internal SRAM question (RP2350 and ESP32-S3)

### DISCUSSION THREAD

**Discussion by AeroSynth** [micropython/micropython#15860] [General]
I had some major problems allocating PSRAM on both _ESP32-S3 WROOM N16R8_ and _Sparkfun Pro Micro RP2350_.  Both boards have 8M PSRAM.
If do this,
```
import gc,array
a=array.array('h',3000000*[0]]
```
it fails to allocate memory. On ESP 32, I can't go beyond 1000000
However, using 
`a=array.array('h',range(3000000))`
works. So using **range** is the right way.  In fact, I can allocate 8M with array.array('h',4000000))
So why does the first method fail?

Another question.  If PSRAM is enabled, is the internal SRAM used to store code or any heap?  Seems internal SRAM would make things run faster.

**Comment by GitHubsSilverBullet** [comment]
`a = array.array('h', 3_000_000*[0]]`

Creates a huge list first.
Then tries to create an additional array and wants to populate it with the list.
Then deletes the list.
Result: uses twice the memory.

`a = array.array('h', range(3_000_000))`
Creates an array and populates it with increasing values from the range iterator.

`a = array.array('h', (0 for _ in range(3_000_000)))`
Creates an array and populates it with 0s.

**Comment by bidrohini** [comment]
When you use 3000000 * [0], Python first creates a list of 3 million elements, each initialized to 0. range(3000000) doesn’t allocate all the memory for the numbers immediately; it’s a generator-like object that yields values one at a time. This method is more memory-efficient and avoids the problem of having to allocate a large temporary list in memory first.
